### DISCUSSION TITLE
What's a good way to make the axis ranges a bit bigger than the data?

### DISCUSSION THREAD

**Discussion by domenicquirl** [GSI-CS-CO/chart-fx#411] [Q&A] [answered]
I'm using an `XYChart`, which automatically very nicely re-adjusts the axes when its data is updated. However, the new axes reach _exactly_ until the last data point, so the last axis label is often cut off:

![image](https://user-images.githubusercontent.com/28535538/116896917-fe521080-ac34-11eb-9a98-1f6683b36487.png)

Or even missing completely:

![image](https://user-images.githubusercontent.com/28535538/116896951-0611b500-ac35-11eb-9c3b-0ea154137fc4.png)

I can think of some ways to manually have my chart track axis bounds and then set them, but there must be a better way. Also, I tried doing a version of that with `XYChart.get[XY]Axis().set(double, double)` when updating data sets to scale the axis by some percentage, and while the new value got set, the visual axis stayed the same. So, how would I best do this? Thanks in advance!

**Comment by wirew0rm** [comment, marked as answer]
Hey, `DefaultNumericAxis`(the default axis implementation) provides `setAutoRangePadding(double)` to which you can e.g. provide `0.1` to add a padding of 10% of the data range to the lower and upper bound of the axis. I'm not sure if it will resolve the incomplete axis labels, as the labels will probably be computed for the range after applying the padding, but it will move the data inwards where there are complete labels.

I believe the labels being cut off is probably a bug. If I remember correctly the canvas for horizontal axes is wider than the plot canvas, to allow the labels to be rendered completely and fit below the vertical axes. Maybe adding a blank axis to the right side of the chart could be a workaround there.

Last but not least you could always override the axis with your own custom implementation to get the maximum amount of control over the displayed range and which tick-marks are generated, but that might be a little bit overkill for your situation.

I hope this helps, please let us know how it works out for you,

Alex

**Reply by domenicquirl** [reply]
Hi Alex,

the auto-padding was exactly what I needed for the data to be more visible! Thanks a lot!

With the padding, because the axis doesn't always range exactly up to a unit tick, it seems to have become reasonably unlikely to hit the case where the label is only partially visible. 

From my side I'm fine with this solution, since the padding is more important to me right now and apparently reduces the label problem by a lot (a custom axis is indeed not necessary for the moment). But how serious are you about the incomplete labels being a bug? Do you want me to open an issue about that (I see you have already gotten around to the one I put up earlier, I'll have a look at that now)?

Cheers,
Domenic

**Reply by wirew0rm** [reply]
Glad it worked for you. We have an internal list of these small inconsistencies that we refer to when we work on the relevant chart subsystems. The problems with things like the axis labels is, that they depend on lots of properties, the data ranges and geometries, so even small changes and bugfixes can have "interesting" effects on seemingly unrelated usecases, so fixing them needs some time for testing. So if you want you can do a proper issue for it, but for us, knowing this problem exists is sufficient.
